File: Lv.1/다트 게임.cpp
#include <string>
#include <stack>
#include <algorithm>
#include <math.h>

using namespace std;

int solution(string dartResult) {
    int answer = 0;
    
    string temp = "";
    stack<int> st;
    
    for(int i=0;i<dartResult.length();i++){
        
        // 숫자가 들어온 경우 -> temp 문자열로 추적
        if(dartResult[i]>='0' && dartResult[i]<='9'){ 
            temp+=dartResult[i];
        }
        
        // 문자 또는 이벤트가 들어온 경우
        else{
            
            // S 점수 계산 후 스택에 넣기
            if(dartResult[i]=='S'){
                
                int cur = stoi(temp);
                temp="";
                cur = pow(cur,1);
                st.push(cur);
                
            }
            
            // D 점수 계산 후 스택에 넣기
            if(dartResult[i]=='D'){
                
                int cur = stoi(temp);
                temp="";
                cur = pow(cur,2);
                st.push(cur);
                
            }
            
            // T 계산 후 스택에 넣기
            if(dartResult[i]=='T'){
                
                int cur = stoi(temp);
                temp="";
                cur = pow(cur,3);
                st.push(cur);
                
            }
            
            // * 연산의 경우
            if(dartResult[i]=='*'){
                // 현재 스택에 1개 있는 경우 하나만 POP 하고 계산 후에 다시 PUSH
                if(st.size()==1){
                    
                    int a = st.top();
                    st.pop();
                    a*=2;
                    st.push(a);
                    
                }
                // 스택에 두개 있는 경우, 둘 다 POP해서 계산 후, 순서에 맞게 다시 넣어줘야한다
                else{
                    int a = st.top();
                    st.pop();
                    int b= st.top();
                    st.pop();
                    
                    a*=2;
                    b*=2;
                    st.push(b);
                    st.push(a);
                }

            }
            
            // # 연산의 경우
            if(dartResult[i]=='#'){
                
                // 현재 거  pop하고 다시 세팅 후 둘 다 push
                int a = st.top();
                st.pop();
                    
                a*=-1;
                st.push(a);
            }
                
            
            
        }
        
    }
    
    // 스택에 있는 모든 값들 더하기
    while(!st.empty()){
        answer+=st.top();
        st.pop();
    }
    return answer;
}

File: Java/LV1/다트 게임.java
import java.util.*;

class Solution {
    public int solution(String dartResult) {
        int answer = 0;
        String temp = "";
        Stack<Integer> st = new Stack<>();
        
        for(int i=0;i<dartResult.length();i++){
            
            // 점수일 경우
            if(dartResult.charAt(i)>='0' && dartResult.charAt(i)<='9'){
                temp+=Character.toString(dartResult.charAt(i));
            }
            
            // 보너스
            if(dartResult.charAt(i)=='S'){
                Integer k = Integer.parseInt(temp);
                k = (int)(Math.pow(k,1));
                st.push(k);
                temp="";
            }
            
            // 보너스
            if(dartResult.charAt(i)=='D'){
                Integer k = Integer.parseInt(temp);
                k = (int)(Math.pow(k,2));
                st.push(k);
                temp="";
            }
            
            // 보너스
            if(dartResult.charAt(i)=='T'){
                Integer k = Integer.parseInt(temp);
                k = (int)(Math.pow(k,3));
                st.push(k);
                temp="";
            }
            
            // 옵션 - 스타상
            if(dartResult.charAt(i)=='*'){
                if(st.size()==1){
                    int a = st.peek();
                    st.pop();
                    a*=2;
                    st.push(a);
                }
                
                if(st.size()==2 || st.size()==3){
                    int a = st.peek();
                    st.pop();
                    int b = st.peek();
                    st.pop();
                    a*=2;
                    b*=2;
                    
                    st.push(b);
                    st.push(a);
                }
                
            }
            
            // 옵션
            if(dartResult.charAt(i)=='#'){
                int a = st.peek();
                st.pop();
                a*=-1;
                st.push(a);
            }
        }
        
        while(!st.isEmpty()){
            int k = st.peek();
            //System.out.println(k);
            st.pop();
            answer+=k;
        }
        return answer;
    }
}
